### PR TITLE
Zetterberg, tanner codes for Xiaozhen Fu

### PR DESCRIPTION
**Pull request by merrysalad**
Thank you for reviewing it.
There is already a file tanner.yml, but I wasn't sure how to update it instead of creating my own copy.

### CODE REVIEW

File: codes/bits/tanner.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  rate: 'For \(C_0\) with rate R, the Tanner code has rate at least 2R-1. For \(C_0\) satisfying Gilbert-Varshamov bound, the rate vs relative distance for the Tanner code is given by \(R \geq 2(1- h(\delta_0))\), where \(\delta_0\) is the relative distance of \(C_0\).'
+
+  encoders:
+    - 'Quadratic algorithm: This technique works for all linear block codes and encodes using matrix multiplication \cite{doi:10.1145/258533.258575}.'
+    - 'Using non-abelian Fast Fourier Transform and exploiting the symmetry of underlying group G, an encoding algorithm that requires \(O(n^{4/3})\) has been devised in \cite{doi:10.1145/258533.258575}.'
+    - 'Spielman modifies the code construction to obtain codes that may be encoded in linear time but give similar performance \cite{10.1109/18.556668}.'
+
+  decoders:
+    - 'Parallel decoding algorithm \cite{doi:10.1109/18.556667} corrects a fraction \(\delta_0^2/48\) of errors.'
+    - 'Decoding algorithm \cite{doi:10.1109/18.910593} gives improved result of correcting a fraction of \(\delta_0/4\) of errors for Tanner codes.'
```
Comment (AlexanderBarg) [inline codes/bits/tanner.yml:29]: \(\delta_0/4\) -> delta_0^2/4

File: codes/bits/zetterberg.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+logical: bits
+
+name: 'Zetterberg code'
+introduced: '\cite{doi:10.1109/TIT.1962.1057682}'
+
+
+description: 'The Zetterberg code \(C_s\) is a binary cyclic code of length \(n = 2^{2s}+1\) generated by the
+minimal polynomial \(g_s(x)\) of \(\alpha\) over GF(2), where \(\alpha\) is a primitive \(n^{th}\) root of unity in the field GF\((2^{4s})\).'
+
+protection: 'Correct all weight-2 errors with a small number of exceptions. By making some alterations to the code, one can form codes that correct all weight-2 errors.'
```
Comment (AlexanderBarg) [inline codes/bits/zetterberg.yml:13]: The code as described has distance 5, and it therefore corrects all weight-2 errors with no exceptions. Not sure which alterations are required.

File: codes/bits/tanner.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#######################################################
```
Comment (valbert4) [inline codes/bits/tanner.yml:1]: The correct directory for the existing tanner.yml file is codes/classical/bits

File: codes/bits/tanner.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    - 'Using non-abelian Fast Fourier Transform and exploiting the symmetry of underlying group G, an encoding algorithm that requires \(O(n^{4/3})\) has been devised in \cite{doi:10.1145/258533.258575}.'
+    - 'Spielman modifies the code construction to obtain codes that may be encoded in linear time but give similar performance \cite{10.1109/18.556668}.'
+
+  decoders:
+    - 'Parallel decoding algorithm \cite{doi:10.1109/18.556667} corrects a fraction \(\delta_0^2/48\) of errors.'
+    - 'Decoding algorithm \cite{doi:10.1109/18.910593} gives improved result of correcting a fraction of \(\delta_0/4\) of errors for Tanner codes.'
+  
+
+notes:
+  - 'Expansion properties of the underlying graphs give efficient decoding algorithms.'
```
Comment (valbert4) [inline codes/bits/tanner.yml:33]: Move expansion statement into description somewhere. Move "If..." statement into protection field.

File: codes/bits/zetterberg.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+realizations:
+  - 'Code used to provide better protection of data transmission with its double error correcting capacity.'
+
+notes:
+  - 'Quasi-perfect, i.e. each n-dimensional binary vector is at most at distance 3 from a codeword.'
+  - 'One of the best known families of double-error correcting binary linear codes.'
+
+relations:
+  parents:
+    - code_id: cyclic
```
Comment (valbert4) [inline codes/bits/zetterberg.yml:37]: also code_id: bits_into_bits

File: codes/bits/zetterberg.yml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+protection: 'Correct all weight-2 errors with a small number of exceptions. By making some alterations to the code, one can form codes that correct all weight-2 errors.'
+
+
+features:
+
+
+  rate: 'The rate is given by \(1-\frac{4s}{n}\), which is asymptotically good, but the minimum distance is a constant of either 5 or 6.'
+
+  encoders:
+    - 'Systematic or non-systematic encoding of information word like in all cyclic codes.'
```
Comment (valbert4) [inline codes/bits/zetterberg.yml:22]: Can remove this.

File: codes/bits/zetterberg.yml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+  decoders:
+    - 'Decoding through linear-feedback shift registers.'
+    - 'Kallquist first described an algebraic decoding theorem. A faster version was later provided in \cite{doi:10.1109/18.149509} and further modified in \cite{doi:10.1109/LCOMM.2010.07.100784} which can detect up to weight error of 3.'
+
+realizations:
+  - 'Code used to provide better protection of data transmission with its double error correcting capacity.'
+
+notes:
+  - 'Quasi-perfect, i.e. each n-dimensional binary vector is at most at distance 3 from a codeword.'
```
Comment (valbert4) [inline codes/bits/zetterberg.yml:32]: Make code_id: perfect a cousin and explain this there. (I've added defn of quasi-perfect in that entry.)

File: codes/bits/zetterberg.yml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  - 'One of the best known families of double-error correcting binary linear codes.'
```
Comment (valbert4) [inline codes/bits/zetterberg.yml:33]: Move to description.

**Pull request closed without merging**